File: app/miscellaneous/config/ConfigReader.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
// SPDX-FileCopyrightText: 2024. The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package miscellaneous.config;

import java.util.List;
import java.util.Map;
import models.user.Role;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;

public interface ConfigReader {
    DateTimeZone getDefaultTimeZone();
    String getHostName();
    Integer getMaxFileSize();
    List<Integer> getExamDurations();
    Integer getExamMaxDuration();
    Integer getExamMinDuration();
    Map<Role, List<String>> getRoleMapping();
    boolean isCourseGradeScaleOverridable();
    boolean isEnrolmentPermissionCheckActive();
    boolean isVisitingExaminationSupported();
    boolean isCollaborationExaminationSupported();
    boolean isHomeExaminationSupported();
    boolean isSebExaminationSupported();
    boolean isCourseSearchActive();
    Map<String, String> getCourseIntegrationUrls();
    DateTime getExamExpirationDate(DateTime timeOfSubmission);
    DateTime getCourseValidityDate(DateTime startDate);
    String getExamExpirationPeriod();
    boolean isMaturitySupported();
    boolean isPrintoutSupported();
    String getAppVersion();
    boolean isAnonymousReviewEnabled();
    String getQuitExaminationLink();
    String getExaminationAdminPassword();
    String getSettingsPasswordEncryptionKey();
    String getHomeOrganisationRef();
    Integer getMaxByodExaminationParticipantCount();
    boolean isByodExamCreationPermissionGrantedForNewUsers();
    String getCourseCodePrefix();
    String getIopHost();
    boolean isApiKeyUsed();
    String getApiKeyName();
    String getApiKeyValue();
    String getPermissionCheckUserIdentifier();
    String getPermissionCheckUrl();
    String getBaseSystemUrl();
    String getSystemAccount();
    String getAttachmentPath();
    String getLoginType();
    String getCsrfCookie();
    boolean isMultiStudentIdEnabled();
    String getMultiStudentOrganisations();
    List<String> getSupportedLanguages();

    boolean hasPath(String path);
    String getString(String path);
}

File: app/services/config/ConfigReader.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.config

import com.google.inject.ImplementedBy
import models.admin.GeneralSettings
import models.user.Role
import org.joda.time.DateTime
import org.joda.time.DateTimeZone

import scala.jdk.CollectionConverters._

@ImplementedBy(classOf[ConfigReaderImpl])
trait ConfigReader:
  def getOrCreateSettings(
      name: String,
      value: Option[String],
      defaultValue: Option[String]
  ): GeneralSettings
  def getDefaultTimeZone: DateTimeZone
  def getHostName: String
  def getMaxFileSize: Int
  def getExamDurations: List[Int]
  def getExamDurationsJava: java.util.List[Integer] = getExamDurations.map(Integer.valueOf).asJava
  def getExamMaxDuration: Int
  def getExamMinDuration: Int
  def getRoleMapping: Map[Role, List[String]]
  def getRoleMappingJava: java.util.Map[Role, java.util.List[String]] =
    getRoleMapping.map((k, v) => k -> v.asJava).asJava
  def isCourseGradeScaleOverridable: Boolean
  def isEnrolmentPermissionCheckActive: Boolean
  def isVisitingExaminationSupported: Boolean
  def isCollaborationExaminationSupported: Boolean
  def isHomeExaminationSupported: Boolean
  def isSebExaminationSupported: Boolean
  def isCourseSearchActive: Boolean
  def getCourseIntegrationUrls: Map[String, String]
  def getExamExpirationDate(timeOfSubmission: DateTime): DateTime
  def getCourseValidityDate(startDate: DateTime): DateTime
  def getExamExpirationPeriod: String
  def isMaturitySupported: Boolean
  def isPrintoutSupported: Boolean
  def getAppVersion: String
  def isAnonymousReviewEnabled: Boolean
  def getQuitExaminationLink: String
  def getExaminationAdminPassword: String
  def getSettingsPasswordEncryptionKey: String
  def getHomeOrganisationRef: String
  def getMaxByodExaminationParticipantCount: Int
  def isByodExamCreationPermissionGrantedForNewUsers: Boolean
  def getCourseCodePrefix: String
  def getIopHost: String
  def isApiKeyUsed: Boolean
  def getApiKeyName: String
  def getApiKeyValue: String
  def getPermissionCheckUserIdentifier: String
  def getPermissionCheckUrl: String
  def getBaseSystemUrl: String
  def getSystemAccount: String
  def getAttachmentPath: String
  def getLoginType: String
  def getCsrfCookie: String
  def isMultiStudentIdEnabled: Boolean
  def getMultiStudentOrganisations: String
  def getSupportedLanguages: List[String]
  def getHomeOrganisations: List[String]
  def isHomeOrganisationRequired: Boolean
  def areNewMultichoiceFeaturesEnabled: Boolean
  def hasPath(path: String): Boolean
  def getString(path: String): String
  def isLocalUser(eppn: String): Boolean
